packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/SerializationTests.kt

```diff
@@ -0,0 +1,446 @@
+            }
+        }
+
+        override fun createPrePopulatedContainer(): SerializableSample =
+            SerializableSample().also {
+                property.get(it)
+                    .apply {
+                        addAll(dataSetToLoad)
+                    }
+            }
+    }
+
+    private class DictionaryTypeSafetyManager<T> constructor(
+        elementType: TypeDescriptor.ElementType,
+        properties: Pair<
+            SerializableSamplePropertyMap<RealmDictionary<Any>>,
+            SerializableSamplePropertyMap<RealmDictionary<Any?>>
+            >,
+    ) : GenericTypeSafetyManager<Pair<String, T>, SerializableSample, RealmDictionary<T>> {
+
+        val classifier: KClassifier = elementType.classifier
+
+        override val dataSetToLoad: List<Pair<String, T>> = getDataSetForDictionaryClassifier(
+            classifier,
+            elementType.nullable,
+            listOf(
+                SerializableSample().apply { stringField = "A" },
+                SerializableSample().apply { stringField = "B" }
+            )
+        )
+
+        override val property = when (elementType.nullable) {
+            false -> properties.first[classifier]!!
+            true -> properties.second[classifier]!!
+        } as KMutableProperty1<SerializableSample, RealmDictionary<T>>
+
+        override fun toString(): String = property.name
+
+        override fun getCollection(container: SerializableSample): RealmDictionary<T> =
+            property.get(container)
+
+        override fun createContainerAndGetCollection(realm: MutableRealm): RealmDictionary<T> {
+            val container = SerializableSample().let {
+                realm.copyToRealm(it)
+            }
+            return property.get(container)
+                .also { dictionary ->
+                    assertNotNull(dictionary)
+                    assertTrue(dictionary.isEmpty())
+                }
+        }
+
+        override fun createPrePopulatedContainer(): SerializableSample {
+            return SerializableSample().also {
+                property.get(it)
+                    .apply {
+                        putAll(dataSetToLoad)
+                    }
+            }
+        }
+    }
+
+    private fun <T> KClassifier.assertValue(
+        expected: T,
+        actual: T
+    ) {
+        when (this) {
+            ByteArray::class -> assertContentEquals(expected as ByteArray?, actual as ByteArray?)
+            RealmObject::class -> assertEquals(
+                (expected as SerializableSample).stringField,
+                (actual as SerializableSample).stringField
+            )
+            RealmAny::class -> {
+                expected as RealmAny?
+                actual as RealmAny?
+
+                if (expected != null && actual != null) {
+                    when (expected.type) {
+                        RealmAny.Type.OBJECT -> {
+                            assertEquals(expected.type, actual.type)
```

Comment:
Couldn't you not just 
```
assertEquals(expected.type, actual.type)
``` 
And then call this method recursively on the value instead of switching on the type?